spec/client/ics-006-solo-machine-client/README.md

```diff
@@ -127,205 +147,178 @@ function latestClientHeight(clientState: ClientState): uint64 {
 }
 ```
 
+### ClientState Methods
+
+All of the functions defined below are methods on the `ClientState` interface. Thus, the solomachine clientstate is always in scope for these functions.
+
 ### Validity predicate
 
-The solo machine client `checkValidityAndUpdateState` function checks that the currently registered public key has signed over the new public key with the correct sequence.
+The solo machine client `verifyClientMessage` function checks that the currently registered public key and diversifier signed over the client message at the expected sequence. If the client message is an update, then it must be the current sequence. If the client message is misbehaviour then it must be the sequence of the misbehaviour.
 
 ```typescript
-function checkValidityAndUpdateState(
-  clientState: ClientState,
-  header: Header) {
-  assert(header.sequence === clientState.consensusState.sequence)
-  assert(header.timestamp >= clientstate.consensusState.timestamp)
-  assert(checkSignature(header.newPublicKey, header.sequence, header.diversifier, header.signature))
-  clientState.consensusState.publicKey = header.newPublicKey
-  clientState.consensusState.diversifier = header.newDiversifier
-  clientState.consensusState.timestamp = header.timestamp
-  clientState.consensusState.sequence++
+function  verifyClientMessage(clientMsg: ClientMessage) {
+  switch typeof(ClientMessage) {
+    case Header:
+      verifyHeader(clientMessage)
+    // misbehaviour only suppported for current public key and diversifier on solomachine
+    case Misbehaviour:
+      verifyMisbehaviour(clientMessage)
+  }
 }
-```
-
-### Misbehaviour predicate
 
-Any duplicate signature on different messages by the current public key freezes a solo machine client.
+function verifyHeader(header: header) {
+    assert(header.timestamp >= clientstate.consensusState.timestamp)
```

Comment:
Should we also `assert(header.sequence === clientState.consensusState.sequence)` as we do [in ibc-go](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/06-solomachine/update.go#L30)?

Reply:
I think the `header.sequence` should be deprecated/removed. I don't believe it is used for anything?

Reply:
Yes, signature can get verified directly against what's in the consensus state so it doesn't have to be included twice just to verify that they are equal.

Deprecated